core/store/src/trie/trie_recording.rs

```diff
@@ -738,3 +786,97 @@ mod trie_recording_tests {
         test_trie_recording_consistency(true, true, true);
     }
 }
+
+#[cfg(test)]
+mod memtrie_batch_iteration_tests {
+    use crate::Trie;
+    use crate::test_utils::{
+        TestTriesBuilder, create_test_store, test_populate_flat_storage, test_populate_trie,
+    };
+    use crate::trie::AccessOptions;
+    use crate::trie::trie_tests::merge_trie_changes;
+    use near_primitives::hash::hash;
+    use near_primitives::shard_layout::ShardUId;
+
+    use super::*;
+
+    /// Returns the hash of height (as le_bytes) for use as a fake block hash in tests.
+    fn fake_hash(height: usize) -> CryptoHash {
+        hash(height.to_le_bytes().as_ref())
+    }
+
+    fn iterate_batch(
+        trie: &Trie,
+        previous_batch_last_key: Option<Vec<u8>>,
+        batch_items: usize,
+    ) -> Option<Vec<u8>> {
+        let read_trie = trie.lock_for_iter();
+        // Get the iterator for the trie, skipping the first key if needed
+        let mut iter = read_trie.iter().expect("failed to get iterator");
+        if let Some(key) = previous_batch_last_key {
+            iter.seek(&key, true).expect("failed to seek");
+        }
+
+        // Iterate over the trie, stopping when we reach the batch size
+        while let Some(result) = iter.next() {
+            let Ok((key, _value)) = result else {
+                panic!("failed to iterate");
+            };
+            // TODO: should we expect here?
+            if trie.recorder_stats().map(|stats| stats.items_count).unwrap_or(0) >= batch_items {
```

Comment:
Why not just use some `counter` variable?

Reply:
we can use a counter variable which is indeed simpler. it would be based on the iterated leafs instead of leafs + internal nodes, which also seems fine.